package pb.ems;

enum BuySell {
  BS_None  = 0;
  BS_Buy   = 1;
  BS_Sell  = 2;
}

enum OrderStatus {
  OS_None       = 0;
  OS_Cancelled  = 1;
  OS_Filled     = 2;
  OS_Rejected   = 3;
  OS_Working    = 4;
  OS_Pending    = 5;
}
  
enum TimeInForce {
  TIF_None  = 0;
  TIF_FAK   = 1;  // allow partial fill
  TIF_FOK   = 2;  // donot allow partial fill
  TIF_GFD   = 3;
  TIF_GTC   = 4;
}

enum OrderType {
  OT_None   = 0;
  OT_Limit  = 1;
  OT_Market = 2;
}

message Status {
  enum Code {
    OK     = 0;
    FAILED = 1;
  }
  optional Code   code = 1;
  optional string msg  = 2;
}

message Order {
  required string               client_order_id      = 1;
  
  optional string               account              = 9;
  optional string               contract             = 10;
  optional BuySell              buy_sell             = 11;
  optional string               price                = 12;
  optional uint64               quantity             = 13;
  optional TimeInForce          tif                  = 14;
  
  optional OrderType            order_type           = 16;

  // updated by cme->client
  optional string      exchange_order_id  = 30;
  optional OrderStatus status             = 31;

  optional string      working_price      = 40;
  optional uint64      working_quantity   = 41;
  optional uint64      filled_quantity    = 42;

  optional string      message            = 50;
  
  optional Timestamp   submit_time        = 60;
}

message Fill {
  required string  fill_id         = 1;
  optional string  fill_price      = 2;
  optional uint64  fill_quantity   = 3;

  optional string    account           = 9;

  optional string    client_order_id   = 10;
  optional string    exchange_order_id = 11;
  optional string    contract          = 12;
  optional BuySell   buy_sell          = 13;
  
  optional Timestamp   fill_time    = 60;
}